### DISCUSSION TITLE
Debug the bibtex item generated by GAP Cite() command.

### DISCUSSION THREAD

**Discussion by hongyi-zhao** [retorquere/zotero-better-bibtex#2510] [General]
This is a continuation of https://github.com/retorquere/zotero-better-bibtex/issues/2504#issuecomment-1543394169.

The BibTeX item is generated as follows:

```shell
$ gap -q <<< "Cite()" |& grep -A 8 '\@misc'
@misc{ GAP4.13dev,
  title =            {{GAP}   {\textendash}   {G}roups,   {A}lgorithms,  and
                      {P}rogramming, {V}ersion 4.13dev},
  organization =     {The GAP {G}roup},
  howpublished =     {\href                     {https://www.gap-system.org}
                      {\texttt{https://www.gap-system.org}}},
  key =              {GAP},
  keywords =         {groups; *; gap; manual}
}
```
Then the year/author fileds are added as follows:

```bibtex
@misc{ GAP4.13dev,
  title =            {{GAP}   {\textendash}   {G}roups,   {A}lgorithms,  and
                      {P}rogramming, {V}ersion 4.13dev},
  organization =     {The GAP {G}roup},
  howpublished =     {\href                     {https://www.gap-system.org}
                      {\texttt{https://www.gap-system.org}}},
  key =              {GAP},
  keywords =         {groups; *; gap; manual},
  year  = {2023},
 author = {The GAP {G}roup}
}
```

Then, the above BibTeX item is copy to clipboard and imported to Zotero, which will generate the following bibtex export:
```bibtex
@misc{GAP4.13dev,
  title = {{{GAP}} \textendash{} {{Groups}}, {{Algorithms}}, and {{Programming}}, {{Version}} 4.13dev},
  author = {Group, The GAP},
  year = {2023},
  publisher = {{The GAP Group}},
  key = {GAP},
  keywords = {*,gap,groups,manual}
}
```

If I remove the author filed, the following bibtex export will be generated:

```bibtex
@misc{GAP4.13dev,
  title = {{{GAP}} \textendash{} {{Groups}}, {{Algorithms}}, and {{Programming}}, {{Version}} 4.13dev},
  year = {2023},
  publisher = {{The GAP Group}},
  key = {GAP},
  keywords = {*,gap,groups,manual}
}
```

Are these the expected results?

Regards,
Zhao

**Comment by hongyi-zhao** [comment]
Below is the reply from [iucr package](https://www.iucr.org/__data/iucr/ftp/templates/latex/) author Brian McMahon <bm@iucr.org>:

> Sorry, I've only just found time to look at this correspondence.
> 
> Dima: Thank you for answering Zhao's queries from your perspective.
> 
> Zhao: Auto-generated bibtex entries are useful, but often need to be
> tweaked to match particular requirements of a publisher. In this case
> I would recommend the following additional tweaks.
> 
> The behaviour of "howpublished" is not well defined
> (https://www.bibtex.com/f/howpublished-field/) and in the current case
> generates an ugly repetition of the web address. Our @misc type
> does support the url field, which I would recommend here.
> 
> Our journal citation style does require a year, which I note that you
> have supplied. If one is not available (or known), just add a field
> like "year = {unknown}".
> 
> Because we use the Harvard referencing style, the author field is
> expected to be something like "Forename Surname" and the style
> transforms it to "Surname, F.". To prevent this, wrap the entire
> author group in another set of braces.
> 
> So I recommend
> 
> @misc{GAP4.13dev,
>    title =  {{GAP} {\textendash} {G}roups, {A}lgorithms, and
>                        {P}rogramming, {V}ersion 4.13dev},
>    organization =     {The GAP {G}roup},
>    url   =     {https://www.gap-system.org/},
>    key =              {GAP},
>    keywords =         {groups; *; gap; manual},
>    year = 2023,
>    author = {{The GAP {G}roup}}
> }
> 
> (In checking this I noticed that there is indeed a bug in iucr.bst
> that leaves a superfluous comma after a "single-name" author. I have
> updated the distributed iucr.bst accordingly and attach a copy for your
> convenience.)
> 
> Regards
> Brian

**Comment by retorquere** [comment]
> which will generate the following bibtex export:
> 
> ```
> @misc{GAP4.13dev,
>   title = {{{GAP}} \textendash{} {{Groups}}, {{Algorithms}}, and {{Programming}}, {{Version}} 4.13dev},
>   author = {Group, The GAP},
>   year = {2023},
>   publisher = {{The GAP Group}},
>   key = {GAP},
>   keywords = {*,gap,groups,manual}
> }
> ```

Correct.

> If I remove the author filed, the following bibtex export will be generated:
> 
> ```
> @misc{GAP4.13dev,
>   title = {{{GAP}} \textendash{} {{Groups}}, {{Algorithms}}, and {{Programming}}, {{Version}} 4.13dev},
>   year = {2023},
>   publisher = {{The GAP Group}},
>   key = {GAP},
>   keywords = {*,gap,groups,manual}
> }
> ```
> 
> Are these the expected results?

Yes.

> So I recommend
> 
> ```
> @misc{GAP4.13dev,
>   title = {{GAP} {\textendash} {G}roups, {A}lgorithms, and {P}rogramming, {V}ersion 4.13dev},
>   organization = {The GAP {G}roup},
>   url = {[https://www.gap-system.org/}](https://www.gap-system.org/%7D),
>   key = {GAP},
>   keywords = {groups; *; gap; manual},
>   year = 2023,
>   author = {{The GAP {G}roup}}
> }
> ```

I agree.

**Reply by retorquere** [reply]
Except the `url` field. That is not valid bibtex.

**Reply by hongyi-zhao** [reply]
> keywords = {groups; *; gap; manual},

Why do the `keywords` need to be adjusted to the above order?

**Reply by retorquere** [reply]
They don't. Keywords are a set, not a list, order is not meaningful. The agreement was wrt how the `author` was entered.

**Reply by hongyi-zhao** [reply]
What does `*;` mean in the keywords?

**Reply by retorquere** [reply]
Nothing. It's in your input `keywords`, so it has been imported as a tag.

**Reply by hongyi-zhao** [reply]
It's not deliberately added by me, but existed in the gap `Cite()` command's output from the very beginning:

```gap
gap> Cite();
Please use one of the following samples
to cite GAP version from this installation

Text:

[GAP]  GAP  –  Groups, Algorithms, and Programming, Version 4.13dev, The GAP
Group, https://www.gap-system.org.

HTML:

<p class='BibEntry'>
[<span class='BibKey'>GAP</span>]   
 <i class='BibTitle'>GAP – Groups, Algorithms,
         and Programming, Version 4.13dev</i>,
 <span class='BibOrganization'>The GAP Group</span>,
<span class='BibHowpublished'><a href="https://www.gap-system.org">https://www\
.gap-system.org</a></span>.
</p>

BibXML:

<entry id="GAP4.13dev"><misc>
  <title><C>GAP</C> &ndash; <C>G</C>roups, <C>A</C>lgorithms,
         and <C>P</C>rogramming, <C>V</C>ersion 4.13dev</title>
  <howpublished><URL>https://www.gap-system.org</URL></howpublished>
  <key>GAP</key>
  <keywords>groups; *; gap; manual</keywords>
  <other type="organization">The GAP <C>G</C>roup</other>
</misc></entry>

BibTeX:

@misc{ GAP4.13dev,
  title =            {{GAP}   {\textendash}   {G}roups,   {A}lgorithms,  and
                      {P}rogramming, {V}ersion 4.13dev},
  organization =     {The GAP {G}roup},
  howpublished =     {\href                     {https://www.gap-system.org}
                      {\texttt{https://www.gap-system.org}}},
  key =              {GAP},
  keywords =         {groups; *; gap; manual}
}

If you are not using BibTeX, here is the bibliography entry produced 
by BibTeX (in bibliography style `alpha'):

\bibitem[GAP]{GAP4}
\emph{GAP -- Groups, Algorithms, and Programming}, Version 4.13dev,
The GAP~Group (today), \verb+https://www.gap-system.org+.

If you have (predominantly) used one or more particular GAP packages,
please cite these packages in addition to GAP itself (either check the
package documentation for the suggestions, or use a scheme like:

[PKG]
<Author name(s)>, <package name>, <package long title>, 
Version <package version> (<package date>), (GAP package),
<package URL>.

You may also produce citation samples for a GAP package by entering

    Cite("packagename");

in a GAP installation with the working version of this package available.
```
Since there is no meaning, why does it produce such output?

**Reply by retorquere** [reply]
> Since there is no meaning, why does it produce such output?

Tags are just free-form labels you can attach to items, and they just don't have any meaning in Zotero. BBT produces this output because you have a tag with value `*`. If you remove that tag, it will no longer show up in the output.

It's not clear to me what you are uncertain about. Titles don't have any meaning either. Whatever you put in the title field will be reproduced in the output. Same goes for keywords. What else would you have wanted BBT to do?

If you're asking "why does GAP put the `*` in the keywords field", you are asking the wrong person. You'll have to ask GAP. If you're offering keywords to import, those get imported as Zotero tags, and in turn, Zotero tags are exported in `keywords`.

**Reply by hongyi-zhao** [reply]
> If you're asking "why does GAP put the * in the keywords field", you are asking the wrong person.

Yes, I mean this.

**Reply by retorquere** [reply]
So you'll have to ask GAP. Before you contacted me I didn't even know GAP existed, I have no idea how they choose their keywords.

**Reply by hongyi-zhao** [reply]
OK, I see.
